pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -472,49 +474,49 @@ func (c *Controller) InstancesByPort(hostname model.Hostname, reqSvcPort int,
 // GetProxyServiceInstances returns service instances co-located with a given proxy
 func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
 	proxyIP := proxy.IPAddress
-	podNamespace := ""
-	if pod, exists := c.pods.getPodByIP(proxyIP); exists {
-		podNamespace = pod.Namespace
+	pod, exists := c.pods.getPodByIP(proxyIP)
+	if !exists {
+		return nil, fmt.Errorf("failed to get pod by proxyIP %s", proxyIP)
 	}
+	podNamespace := pod.Namespace
+
+	out := make([]*model.ServiceInstance, 0)
+
+	// 1. find proxy service by label selector, if not any, there may exist headless service
+	// failover to 2
+	svcLister := listerv1.NewServiceLister(c.services.informer.GetIndexer())
```

Comment:
For the second part of the optimization: I would point that we now have a 'workload' callback, which creates an index in the ADS server containing all workloads and their labels and endpoints.

It would be ideal to move this optimization there - and create a secondary index by IP ( if it doesn't exist ). 

Longer term we may stop using the informer for pods ( which holds in memory the full info about a pod - can be quite large ), and use a simple watcher instead.

Reply:
>For the second part of the optimization: I would point that we now have a 'workload' callback, which creates an index in the ADS server containing all workloads and their labels and endpoints.

This is reasonable

>Longer term we may stop using the informer for pods ( which holds in memory the full info about a pod - can be quite large ), and use a simple watcher instead.

I am not quite sure about what you mean by `use a simple watcher`, as long as we watch resources from kube-apiserver, the client will holds in memory all the objects. This is decided by client-go implementation.